source/Scope.cpp

```diff
@@ -4,18 +4,42 @@
 namespace slang {
 
 static const Scope emptyScope;
-const Scope* const Scope::Empty = &emptyScope;
+const Scope *const Scope::Empty = &emptyScope;
 
-void Scope::add(const Symbol* symbol) {
-    ASSERT(symbol);
-    table.emplace(symbol->name, symbol);
+Scope::Scope() : parentScope{Scope::Empty} {}
+
+bool Scope::add(const Symbol *symbol) {
+        ASSERT(symbol);
+    list.emplace_back(symbol);
+    return table.emplace(symbol->name, symbol).second;
+}
+
+const Symbol *Scope::lookup(StringRef name, bool local) const {
+    for (auto scope = this; scope && scope != Scope::Empty; scope = scope->parent()) {
```

Comment:
so Scope::Empty is just an optimization instead of default-constructed Scope() ?